lib/cdo/aws/resource_management/ec2_instance_for_deletion.rb

```diff
@@ -0,0 +1,67 @@
+# EC2InstanceForDeletion class controls the deletion process of an EC2 instance and provides methods to evaluate
+# if an instance is eligible for deletion.
+#
+# It includes functionalities to check if the API termination is disabled, if the instance has an associated
+# stack for deletion, if the stack is critical, and if the instance has been stopped for a minimum duration.
+#
+# Additionally, it retrieves necessary information such as the stack name and stop time, allowing users to
+# determine the eligibility of an EC2 instance for deletion based on defined criteria.
+class EC2InstanceForDeletion
+  # Constants defining tags and critical stacks
+  DISABLED_API_TERMINATION_TAG = 'disableApiTermination'.freeze
+  CLOUD_FORMATION_TAG = 'aws:cloudformation:stack-name'.freeze
+  CRITICAL_STACKS = %w[autoscale-prod test staging levelbuilder adhoc-bugcrowd]
+
+  # Initialize EC2 instance and optional deletion configuration options
+  def initialize(instance, opts = {})
+    @instance = instance
+    @deletion_tag = opts[:deletion_tag]
+    @minimum_time_stopped_in_seconds = opts.fetch(:minimum_time_stopped_in_seconds, 6 * 24 * 60 * 60) # Default 6 days
+    @_dry_run = opts[:dry_run]
+  end
+
+  # Check if the EC2 instance is eligible for deletion
+  def can_be_deleted?
+    !api_termination_disabled? && has_stack_to_delete? && !has_critical_stack? && has_reached_minimum_stop_time?
+  end
```

Comment:
I don't love how many tiny private sub-methods which only ever get used once we're using here. What about instead something like:

```suggestion
  def can_be_deleted?
    # comment explaining what this tag is and why we use it
    return false unless @instance.describe_attribute(attribute: DISABLED_API_TERMINATION_TAG).disable_api_termination.value
    
    # comment explaining the scenario in which it's possible to not have a stack name
    return false if stack_name.nil?
    
    # comment explaining what it means for a stack to be critical
    return false if CRITICAL_STACKS.include?(@instance.stack_name)
    
    # probably don't need much of a comment here, as long as the has_reached_minimum_stop_time method is well-commented
    return false if has_reached_minimum_stop_time?
    
    return true
  end
```

Reply:
What about a middle ground? Something like,

```suggestion
  def can_be_deleted?
    return false if stack_name.nil?
    return false if CRITICAL_STACKS.include?(@instance.stack_name)
    return false if api_termination_disabled?
    return false unless has_reached_minimum_stop_time?
    return true
  end
```

Feels like it strikes a good balance between readability and encapsulation. Also makes it easier to notice something odd which I hadn't noticed on first review; why do we use `@instance.stack_name` in one place, and `@instance.tags.find {|tag| tag.key == CLOUD_FORMATION_TAG}&.value` elsewhere?